pkg/controller/jobs/sparkapplication/sparkapplication_controller.go

```diff
@@ -0,0 +1,320 @@
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloads,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloads/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloads/finalizers,verbs=update
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=resourceflavors,verbs=get;list;watch
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloadpriorityclasses,verbs=get;list;watch
+
+func NewJob() jobframework.GenericJob {
+	return &SparkApplication{}
+}
+
+var NewReconciler = jobframework.NewGenericReconcilerFactory(NewJob)
+
+type SparkApplication sparkv1beta2.SparkApplication
+
+var _ jobframework.GenericJob = (*SparkApplication)(nil)
+
+func (j *SparkApplication) Object() client.Object {
+	return (*sparkv1beta2.SparkApplication)(j)
+}
+
+func (j *SparkApplication) IsSuspended() bool {
+	return j.Spec.Suspend != nil && *j.Spec.Suspend
+}
+
+func (j *SparkApplication) IsActive() bool {
+	return j.Status.AppState.State == sparkv1beta2.ApplicationStateRunning
+}
+
+func (j *SparkApplication) Suspend() {
+	j.Spec.Suspend = ptr.To(true)
+}
+
+func (j *SparkApplication) GVK() schema.GroupVersionKind {
+	return gvk
+}
+
+func (j *SparkApplication) PodLabelSelector() string {
+	return fmt.Sprintf("%s=%s", sparkcommon.LabelSparkAppName, j.Name)
+}
+
+func (j *SparkApplication) PodSets(ctx context.Context) ([]kueue.PodSet, error) {
+	// driver and executor
+	podSets := make([]kueue.PodSet, 2)
+
+	// driver
+	driverPodTemplateSpec, err := j.buildDriverPodTemplateSpec()
+	if err != nil {
+		return nil, err
+	}
+	podSets[0] = kueue.PodSet{
+		Name:     driverPodSetName,
+		Template: *driverPodTemplateSpec,
+		Count:    1,
+	}
+
+	if features.Enabled(features.TopologyAwareScheduling) {
+		topologyRequest, err := jobframework.NewPodSetTopologyRequest(
+			&driverPodTemplateSpec.ObjectMeta).Build()
+		if err != nil {
+			return nil, err
+		}
+		podSets[0].TopologyRequest = topologyRequest
+	}
+
+	// executors
+	executorPodTemplateSpec, err := j.buildExecutorPodTemplateSpec()
+	if err != nil {
+		return nil, err
+	}
+	podSets[1] = kueue.PodSet{
+		Name:     executorPodSetName,
+		Template: *executorPodTemplateSpec,
+		Count:    j.numInitialExecutors(),
+	}
+
+	if features.Enabled(features.TopologyAwareScheduling) {
+		topologyRequest, err := jobframework.NewPodSetTopologyRequest(
+			&executorPodTemplateSpec.ObjectMeta).Build()
+		if err != nil {
+			return nil, err
+		}
+		podSets[1].TopologyRequest = topologyRequest
+	}
+
+	return podSets, nil
+}
+
+func (j *SparkApplication) RunWithPodSetsInfo(ctx context.Context, podSetsInfo []podset.PodSetInfo) error {
+	expectedLen := 2 // driver + executor
+	if len(podSetsInfo) != expectedLen {
+		return podset.BadPodSetsInfoLenError(expectedLen, len(podSetsInfo))
+	}
+
+	j.Spec.Suspend = ptr.To(false)
+
+	// driver
```

Comment:
It would be nice this block of code into a function.

It looks to be the same just differing on driver or executor.